compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java

```diff
@@ -2025,11 +2030,24 @@ private BType defineSymbolForCyclicTypeDefinition(BLangTypeDefinition typeDef, S
                         newTypeDefName, env.enclPkg.symbol.pkgID, newTypeNode, env.scope.owner,
                         typeDef.name.pos, SOURCE);
                 break;
```

Comment:
I'm not sure this is the correct fix. 

The following doesn't panic atm.

```ballerina
type T readonly & [T];

public function main() {
    [error] x = [error("")];
    T t = <T> x;
}
```

Same for immutable list values that are not recursive.

Reply:
We need to make sure the type is correct and add tests for the same, both at compile-time and runtime.

Reply:
Similarly, no compilation error atm for

```ballerina
type T readonly & [T];

public function main() {
    [int] x = [1];
    T t = x;
}
```

Reply:
Currently this PR mostly focus about the scenarios when recursive type is defining.
It is not consider assignments.
Shall I create a new issue and work separately or Shall I add the changes in same PR?.